src/if_statements.py

```diff
@@ -600,27 +648,56 @@ def emit_return(context: Context, node: ReturnNode, body: Body) -> None:
 
 def build_switch_between(
     context: Context,
-    start: SwitchNode,
+    switch: SwitchNode,
     end: Node,
 ) -> SwitchStatement:
-    """Output the subgraph between `start` and `end`, including the jump
-    in the SwitchStatement `start`, but not including `end`"""
-    body = Body(print_node_comment=context.options.debug)
-    jump = switch_jump(context, start)
-    emitted_cases: Set[Node] = set()
-
-    for case in start.cases:
-        if case in context.emitted_nodes:
-            continue
-        if case == end:
-            # We are not responsible for emitting `end`
-            continue
+    """
+    Output the subgraph between `switch` and `end`, but not including `end`.
+    The returned SwitchStatement starts with the jump to the switch's value.
+    """
+    # Try to detect the default node by looking at the switch's parent
+    default_node: Optional[Node] = None
+    if (
+        len(switch.parents) == 1
+        and isinstance(switch.parents[0], ConditionalNode)
+        and switch.parents[0].fallthrough_edge is switch
+        and switch.parents[0].conditional_edge is not end
+    ):
+        default_node = switch.parents[0].conditional_edge
+
+    add_labels_for_switch(context, switch, default_node)
+    switch_index = context.switch_nodes[switch]
+    comment = f" // switch {switch_index}" if switch_index else ""
 
-        body.extend(build_flowgraph_between(context, case, end))
-        # TODO: This could be a `break;` statement
-        if not body.ends_in_jump():
-            emit_goto_or_early_return(context, end, body)
-    return SwitchStatement(jump, body)
+    jump: Optional[Union[Expression, SimpleStatement]]
+    jump = get_block_info(switch).switch_value
+    if not isinstance(jump, SwitchControl):
+        jump = switch_jump(context, switch)
+
+    switch_body = Body(print_node_comment=context.options.debug)
+
+    # Order case blocks by their position in the asm, not by their order in the jump table
+    sorted_cases = sorted(set(switch.cases), key=lambda node: node.block.index)
+    next_sorted_cases: List[Optional[Node]] = []
+    next_sorted_cases.extend(sorted_cases[1:])
+    next_sorted_cases.append(None)
+    for case, next_case in zip(sorted_cases, next_sorted_cases):
+        if case in context.emitted_nodes or case is end:
+            pass
+        elif (
+            next_case is not None
+            and next_case not in context.emitted_nodes
```

Comment:
why is this check here? If I understand correctly it's harmless, because the `else` codepath is also correct, just with missing `// fallthrough` comments, but it feels off to me

Reply:
If `next_case` is already emitted, then nothing will be emitted in the next iteration (to avoid duplicating the node). This is an unusual situation - this means that `next_node` has a parent outside of the `SwitchNode` tree.

Commenting out this check results in this diff in `multi-switch`. Here, node 16 can be reached from either switch 1 or 2 (and was originally emitted in switch 1). The "fallthrough" comment isn't correct here.
```diff
--- 

+++ 

@@ -74,8 +74,7 @@

                     return arg0 * arg0;
                 case 2: // switch 2
                     phi_a0_2 = arg0 - 1;
-                    // Duplicate return node #16. Try simplifying control flow for better match
-                    return (phi_a0_2 + 1) ^ phi_a0_2;
+                    // fallthrough
                 case 6: // switch 2
                 case 7: // switch 2
                     phi_a0_3 = arg0 * 2;
```